From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aleksei Kuzmin <alkuzmin@microsoft.com>
Date: Mon, 19 Nov 2018 16:35:11 +0100
Subject: build: register Microsoft Node.js modules

Microsoft patch

diff --git a/shell/common/node_bindings.cc b/shell/common/node_bindings.cc
index 10b20f68326239078c870237590601fd78c389ba..0ac3a0a41d320e5e6688b069e9b219aac95f43d2 100644
--- a/shell/common/node_bindings.cc
+++ b/shell/common/node_bindings.cc
@@ -24,6 +24,7 @@
 #include "content/public/browser/browser_thread.h"
 #include "content/public/common/content_paths.h"
 #include "electron/buildflags/buildflags.h"
+#include "microsoft/src/electron/shell/node_bindings_ms.h"
 #include "shell/common/electron_command_line.h"
 #include "shell/common/gin_converters/file_path_converter.h"
 #include "shell/common/gin_helper/dictionary.h"
@@ -329,6 +330,7 @@ void NodeBindings::RegisterBuiltinModules() {
   ELECTRON_DESKTOP_CAPTURER_MODULE(V)
 #endif
 #undef V
+  microsoft::RegisterBuiltinModules();
 }
 
 bool NodeBindings::IsInitialized() {
