From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aleksei Kuzmin <alkuzmin@microsoft.com>
Date: Mon, 12 Nov 2018 19:05:45 +0100
Subject: feat: add "microsoft-build" to the process.versions dict

Microsoft patch

diff --git a/shell/common/api/electron_bindings.cc b/shell/common/api/electron_bindings.cc
index 5d1bade086bbfeffabaa1ed089af07ad352d47dc..2e1b347a0922174b617e7608b5fda8f163445297 100644
--- a/shell/common/api/electron_bindings.cc
+++ b/shell/common/api/electron_bindings.cc
@@ -123,6 +123,9 @@ void ElectronBindings::BindTo(v8::Isolate* isolate,
   gin_helper::Dictionary versions;
   if (dict.Get("versions", &versions)) {
     versions.SetReadOnly(ELECTRON_PROJECT_NAME, ELECTRON_VERSION_STRING);
+#if BUILDFLAG(MICROSOFT_ADD_BUILD_VERSION_TO_PROCESS_VERSIONS)
+    versions.SetReadOnly("microsoft-build", MICROSOFT_BUILD_VERSION_STRING);
+#endif
     versions.SetReadOnly("chrome", CHROME_VERSION_STRING);
   }
 }
